spacy/language.py

```diff
@@ -1827,6 +1830,12 @@ def from_config(
                     source_name = pipe_cfg.get("component", pipe_name)
                     listeners_replaced = False
                     if "replace_listeners" in pipe_cfg:
+                        # HACK: Reset any listened-to components to the listener
+                        # state of the source pipeline for the purpose of
+                        # replacing listeners. The add_pipe below will set
+                        # the state back to the listener state for the new
+                        # pipeline.
+                        source_nlps[model]._link_components()
```

Comment:
It's nice to be able to remove the whole extra block L1895-L1915, but this hack is giving me a headache. I've been trying to think of ways on how we could do this more cleanly instead... Because right now it feels a bit obscure and difficult to maintain in the future...

Reply:
Does this PR need the rewrite from the original L1895-L1915 block to this `source_nlps[model]._link_components()` call? It seems like the tests are passing without this edit?

Reply:
I do think we need to clean this up, but it feels like a separate thing and we probably need to write more unit tests for it, too?

Reply:
You need the source pipeline to have the right listener state or `replace_listeners` (in particular from `spacy-transformers`) does not work correctly. Use this test case: https://github.com/explosion/spacy-transformers/pull/381

Reply:
Ah, wait, I may have this backwards again. Let me send you a test case separately, but it's mainly related to the `replace_listeners` code in `spacy-transformers`, not something that's easy to test in the spacy test suite.

Reply:
The code in L1895-L1915 now always does nothing because the listener state in the new pipeline has already been configured correctly.

Reply:
If by "now" you mean "in this PR", I get that. 

My question is: what is a test case that fails with the original block in place vs. the one-liner in this PR? Because if I switch them back, I don't get any failing tests (also not in the `spacy-transformers` PR you linked). It would be helpful to understand the refactor & also to have the test suite cover that.

Reply:
Yeah, it looks like there are no tests for `replace_listener` and `replace_listener_cfg` outside of `spacy-transformers`, and `spacy-transformers` tests don't cover this particular case either because you need a case where multiple components and the `transformer` component are sourced at the same time.